scheds/rust/scx_mitosis/src/bpf/mitosis.bpf.c

```diff
@@ -0,0 +1,932 @@
+	return NULL;
+}
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+	__type(key, u32);
+	__type(value, struct cpu_ctx);
+	__uint(max_entries, 1);
+} cpu_ctxs SEC(".maps");
+
+static inline struct cpu_ctx *lookup_cpu_ctx(int cpu)
+{
+	struct cpu_ctx *cctx;
+	u32 zero = 0;
+
+	if (cpu < 0)
+		cctx = bpf_map_lookup_elem(&cpu_ctxs, &zero);
+	else
+		cctx = bpf_map_lookup_percpu_elem(&cpu_ctxs, &zero, cpu);
+
+	if (!cctx) {
+		scx_bpf_error("no cpu_ctx for cpu %d", cpu);
+		return NULL;
+	}
+
+	return cctx;
+}
+
+struct cell {
+	u64 vtime_now;
+	u32 dsq;
+	// The following field is populated from userspace to indicate
+	// which cpus the cell should belong to.
+	unsigned char cpus[MAX_CPUS_U8];
+};
+
+struct cell cells[MAX_CELLS];
+
+static inline struct cell *lookup_cell(int idx)
+{
+	if (idx < 0 || idx >= MAX_CELLS) {
+		scx_bpf_error("invalid cell %d", idx);
+		return NULL;
+	}
+	return &cells[idx];
+}
+
+/*
+ * Store the cpumask for each cell (owned by BPF logic)
+ */
+struct cell_cpumask_wrapper {
+	struct bpf_cpumask __kptr *cpumask;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_ARRAY);
+	__type(key, u32);
+	__type(value, struct cell_cpumask_wrapper);
+	__uint(max_entries, MAX_CELLS);
+	__uint(map_flags, 0);
+} cell_cpumasks SEC(".maps");
+
+static inline struct cpumask *lookup_cell_cpumask(int idx)
+{
+	struct cell_cpumask_wrapper *cpumaskw;
+
+	if (!(cpumaskw = bpf_map_lookup_elem(&cell_cpumasks, &idx))) {
+		scx_bpf_error("no cell cpumask");
+		return NULL;
+	}
+
+	return (struct cpumask *)cpumaskw->cpumask;
+}
+
+/* css iteration is experimental, forward declare the necessary functions */
+struct bpf_iter_css;
+struct cgroup_subsys_state;
+extern int bpf_iter_css_new(struct bpf_iter_css *it,
+			    struct cgroup_subsys_state *start,
+			    unsigned int flags) __weak __ksym;
+extern struct cgroup_subsys_state *
+bpf_iter_css_next(struct bpf_iter_css *it) __weak __ksym;
+extern void bpf_iter_css_destroy(struct bpf_iter_css *it) __weak __ksym;
+
+/*
+* Along with a user_global_seq bump, indicates that cgroup->cell assignment
+* changed
+*/
+volatile bool update_cell_assignment;
+bool draining;
+
+/*
+ * This is the main driver for reconfiguration. It only runs on CPU 0
+ * To avoid
+ */
+SEC("fentry/scheduler_tick")
+int scheduler_tick_fentry(const void *ctx)
+{
+	int cell_idx, cpu_idx;
+	struct cpu_ctx *cpu_ctx;
+	struct bpf_cpumask *cpumask;
+	struct cell_cpumask_wrapper *cell_cpumaskw;
+	struct cgroup_subsys_state *root_css, *pos;
+	struct cgroup *root_cgrp;
+
+	if (bpf_get_smp_processor_id() != 0)
+		return 0;
+
+	/*
+	 * To handle races where tasks are assigned to cells that are getting
+	 * removed, we ensure cpus dispatch from their previous cell for an entire
+	 * scheduler tick. This is a crude way of mimicing RCU synchronization.
+	 */
+	if (draining) {
+		bpf_for(cpu_idx, 0, nr_possible_cpus)
+		{
+			if (!(cpu_ctx = lookup_cpu_ctx(cpu_idx)))
+				return 0;
+
+			cpu_ctx->prev_cell = cpu_ctx->cell;
+		}
+		barrier();
```

Comment:
> it can be emulated with a sync RMW op on a dummy var.

Do you have an example? Using atomics typically results in llvm crashes for me. This particular barrier is just because I have a debug check in select_cpu that wants to confirm if we're not draining, then we don't misfire a task to the wrong cell. Otherwise I only check draining on tick()

Reply:
`__sync_fetch_and_add()` on a dummy variable (percpu if perf is a concern) should do. It's used quite a bit in C scheds for stats and other stuff, so should be okay in terms of compiler support. It's just that `barrier()` doesn't really do much when multiple CPUs need to be interlocked as the memory accesses can be reordered across it, so `select_cpu()` may see the updates in the reverse order regardless of `barrier()`.

Reply:
What does that __sync_fetch_and_add() do in that case?

To be clear, this is all a bit academic on x86, right? My understanding is that since x86 has a TSO memory model, the compiler barrier is all I need to ensure the write to prev_cell is always seen before the write to draining

Reply:
Atomic RMW implies full memory barrier, so we're just using it for the barrier part.

On x86, the stores are always full ordered but loads aren't, so `wmb()` is the same as `barrier()` but `rmb()` actually needs to tell the CPU that things can't be reordered across it. So, it'll misbehave on x86 too.

Reply:
Would I need the atomic RMW on both the write and read side?

Reply:
That usually is the case. Maybe I'm not understanding the code correctly but it's usually the following pattern:
```
           A                 C
         mb();              mb();
           B                 D
```
Where B looks at what C did and D looks at what A did. The guarantee that's created by the paired `mb()`s is that either B sees C or D sees A.

Reply:
All I care about is the following sequence:

```
write A                       read B
write B                       read A
```
I need to ensure that if read B sees write B then the read A *must* see write A. My understanding is that on x86, `mov` has these semantics. I just need to prevent the compiler from reordering the write (or read). Adding an RMW on both sides is super heavyweight for this.

Reply:
Yeah, on x86, wmb() is the same as barrier(), so it'd only need barrier() on the writer side. Isn't reader side hotter tho? Would it make difference whether the writer side has one more atomic op or not?